include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,873 @@
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_comp_session_create(odp_comp_session_param_t *param,
+			    odp_comp_session_t *session,
+			    odp_comp_ses_create_err_t *status);
+
+/**
+ * Comp session destroy
+ *
+ * Destroy an unused session. Result is undefined if session is being used
+ * (i.e. asynchronous operation is in progress).
+ *
+ * @param session           Session handle
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_comp_session_destroy(odp_comp_session_t session);
+
+/**
+ * Comp set dictionary
+ *
+ * Should be called when there is no operation in progress i.e.
+ * before initiating processing of first chunk of data and
+ * after processing of last chunk of data is complete.
+ *
+ * @param session           Session handle
+ * @param dict[in,out]      Pointer to dictionary.
+ *                          implementation should update length of dictionary
+ *                          used at output.
+ * @retval 0 on success
+ * @retval <0 on failure
+ *
+ * @note:
+ * Application should call odp_comp_alg_capability() to query 'support_dict'
+ * before making call to this API.
+ */
+int odp_comp_set_dict(odp_comp_session_t session,
+		      odp_comp_dict_t *dict);
+
+/**
+ * Comp API to process a packet synchronously
+ *
+ * If session is created in ODP_COMP_SYNC mode, this call wait for operation
+ * to complete.
+ *
+ * If session is created in ODP_COMP_ASYNC mode, behavior is undefined.
+ *
+ * For compression + hash, call returns with hash appended at the end of
+ * packet which has last field set in corresponding param.
+ *
+ * User should compute processed data len = output_data_range.len - digest_len,
+ * where digest_len queried through odp_comp_hash_alg_capability().
+ *
+ * For every input packet, user should pass corresponding valid output
+ * packet.
+ *
+ * For every successful or partially processed packet, api update result param
+ * with out_data_range indicating length of the data written in output packet
+ * and error code indicating operation status.
+ *
+ * A packet is said to be partially processed if it runs out of buffer during
+ * processing and need more buffer from user to continue. Under such condition,
+ * API fails with error code code in result param set to
+ * ODP_COMP_ERR_OUT_OF_SPACE.
+ *
+ * In case of ODP_COMP_ERR_OUT_OF_SPACE error, application should
+ * keep on calling odp_comp_op() API with more output buffer until call returns
+ * with success or any other error code.
+ *
+ * @param         pkt_in   Input Packet
+ * @param         pkt_out  Output Packet
+ * @param         param    Operation parameter
+ * @param[out] result   Result parameter, containing after operation result.
+ *
+ * @return 0 on success
+ * @retval <0 on failure
+ */
+int odp_comp_op(const odp_packet_t pkt_in,
+		const odp_packet_t pkt_out,
+		const odp_comp_packet_op_param_t *param,
+		odp_comp_packet_op_result_t *result);
+
```

Comment:
Application does not allocate buffers but packets.

Does the partially processed packet contain valid data or not == does application keep or discard the packet?

Is there any way for application to prepare large enough packet in advance? E.g. a per algo capability of worst case decompression ratio? Should "pkt_out = INVALID" be supported? ==> implementation allocates large enough packet.

Single packet handle should be defined as const.

"param" is [in, out] since it contains ranges

Reply:
Does the partially processed packet contain valid data or not == does application keep or discard the packet?
>> It will be a valid data and thus application is supposed to consume it. 
Is there any way for application to prepare large enough packet in advance?
>> Its more of application dependent such as ipcomp it knows packet cannot go beyond a maximum length of IP Packet.
E.g. a per algo capability of worst case decompression ratio?
>> No. it is not guaranteed. Its more of data type property and algorithm capability. 
Should "pkt_out = INVALID" be supported? 
>> No. 
implementation allocates large enough packet.
>> we refrained from this approach as in such case implementation may end up eating whole of user pool. Better approach is pass back what ever is available at output. Now, application may consume it and reuse same packet for next o/p OR extend the packet by some length (say double it up).
Single packet handle should be defined as const.
>> It is const in API param list ... odp_comp_op(const odp_packet_t pkt_in....) .. what am i missing here?
There is out_data_range defined in both op_param_t and op_result_t. Which one is updated on output? 
>> One in op_result.
May be operation output comments should be removed from op_param_t documentation.
>> Yea. Did that already.

Please mark your approval to these.

Reply:
It looks like OUT_OF_SPACE causes too much controversity ATM. I would suggest to make it a hard error at this point, merge API w/o out-of-space, then rethink this particular item. Partially processed packets need additional thought with respect to state being saved/passed back/etc.

Reply:
Single decompress (or any other) operation should not eat up entire pool, but try to allocate one packet that is large enough (up to the max packet size of the pool).

Agree, OUT_OF_SPACE is problematic. It's kind of optimization for minimizing (output packet) memory usage. If implementation would allocate output packet, it would be optimal in ease-of-use, performance and memory usage.